source/llassetgen/source/Image.cpp

```diff
@@ -0,0 +1,358 @@
+	template LLASSETGEN_API void Image::setPixel<uint32_t>(const Vec2<size_t> pos, const uint32_t in) const;
+	template LLASSETGEN_API void Image::setPixel<uint16_t>(const Vec2<size_t> pos, const uint16_t in) const;
+	template LLASSETGEN_API void Image::setPixel<uint8_t>(const Vec2<size_t> pos, const uint8_t in) const;
+
+	template <typename pixelType>
+	void Image::exportPng(const std::string &filepath, pixelType black, pixelType white) {
+		std::ofstream out_file(filepath, std::ofstream::out | std::ofstream::binary);
+		if (!out_file.good()) {
+			std::cerr << "could not open file " << filepath;
+			abort();
+		}
+
+		png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
+		if (!png) {
+			std::cerr << "failed to create png write struct";
+			abort();
+		}
+
+		png_infop info = png_create_info_struct(png);
+		if (!info) {
+			png_destroy_write_struct(&png, (png_infopp)0);
+			std::cerr << "failed to create png info struct";
+			abort();
+		}
+
+		if (setjmp(png_jmpbuf(png))) {
+			std::cerr << "pnglib caused a longjump due to an error" << std::endl;
+			std::cerr << "could not write file " << filepath;
+			abort();
+		}
+
+		png_set_write_fn(png, (png_voidp)&out_file, write_data, flush_data);
+
+		png_set_IHDR(png,
+			info,
+			get_width(),
+			get_height(),
+			(bit_depth < 16) ? bit_depth : 16,
+			PNG_COLOR_TYPE_GRAY,
+			PNG_INTERLACE_NONE,
+			PNG_COMPRESSION_TYPE_BASE,
+			PNG_FILTER_TYPE_BASE);
+
+		png_write_info(png, info);
+
+		if (bit_depth > 8) {
+			png_set_swap(png);
+		}
+
+		if (bit_depth >= 24) {
+			std::unique_ptr<uint16_t[]> row(new uint16_t[get_width()]);
+			// possible 32 float or 32 or 24 bit int data
+			// scale down to 16 bit int grayscale
+
+			for (size_t y = 0; y < get_height(); y++) {
+				for (size_t x = 0; x < get_width(); x++) {
+                    row[x] = clamp(static_cast<float>(getPixel<pixelType>({x, y}) - black) / (white - black), 0.0F, 1.0F) * std::numeric_limits<uint16_t>::max();
+				}
+				png_write_row(png, reinterpret_cast<png_bytep>(row.get()));
+			}
+		} else {
+			for (size_t y = 0; y < get_height(); y++) {
+				png_write_row(png, reinterpret_cast<png_bytep>(&data.get()[y * stride]));
+			}
+		}
+
+		png_write_end(png, nullptr);
+
+		png_destroy_write_struct(&png, &info);
+		out_file.close();
+	}
+
+	template LLASSETGEN_API void Image::exportPng<uint32_t>(const std::string &filepath, uint32_t min, uint32_t max);
+	template LLASSETGEN_API void Image::exportPng<uint16_t>(const std::string &filepath, uint16_t min, uint16_t max);
+	template LLASSETGEN_API void Image::exportPng<uint8_t>(const std::string &filepath, uint8_t min, uint8_t max);
+	template LLASSETGEN_API void Image::exportPng<float>(const std::string &filepath, float min, float max);
+
+	void Image::read_data(png_structp png, png_bytep data, png_size_t length) {
+		png_voidp a = png_get_io_ptr(png);
+		((std::istream*)a)->read((char*)data, length);
```

Comment:
I suggest to add an assertion that a `dynamic_cast` to `std::istream*` results in a valid pointer.